projects/eslint-config/plugins/portal/lib/rules/no-react-dom-create-portal.js

```diff
@@ -0,0 +1,213 @@
+			const ancestors = context.getAncestors(node);
+
+			const parent = ancestors[ancestors.length - 1];
+
+			return (
+				parent.type === 'ImportDeclaration' &&
+				parent.source.type === 'Literal' &&
+				parent.source.value === 'react-dom'
+			);
+		};
+
+		// Returns true if `identifier` actually refers to
+		// `variable`, by walking up the scope stack.
+
+		const isSame = (identifier, variable) => {
+			const name = variable.name;
+			let scope = context.getScope(identifier);
+
+			while (scope) {
+				for (let i = 0; i < scope.variables.length; i++) {
+					if (scope.variables[i] === variable) {
+						return true;
+					}
+					else if (scope.variables[i].name === name) {
+
+						// Variable is shadowed, so it is not the same.
+
+						return false;
+					}
+				}
+				scope = scope.upper;
+			}
+		};
+
+		const report = (node) =>
+			context.report({
+				messageId: 'noReactDOMCreatePortal',
+				node,
+			});
+
+		const foundBindings = new Set();
+		const foundNamespaces = new Set();
+
+		const add = (set, iterable) => {
+			for (const item of iterable) {
+				set.add(item);
+			}
+		};
+
+		return {
+			CallExpression(node) {
+				if (
+					node.callee.type === 'MemberExpression' &&
+					node.callee.object.type === 'Identifier'
+				) {
+
+					// eg. Foo.bar()
+
+					for (const namespace of foundNamespaces) {
+						if (
+							node.callee.object.name === namespace.name &&
+							node.callee.property.type === 'Identifier' &&
+							node.callee.property.name === 'createPortal'
+						) {
+
+							// eg. Foo.createPortal()
+
+							if (isSame(node.callee.object, namespace)) {
+
+								// "Foo" came from react-dom.
+
+								report(node);
+							}
+						}
+					}
+				}
+				else if (node.callee.type === 'Identifier') {
+
+					// eg. something()
+
+					for (const binding of foundBindings) {
+						if (node.callee.name === binding.name) {
+
+							// "something" came from react-dom.
+
+							if (isSame(node, binding)) {
+								report(node);
+							}
+						}
+					}
+				}
+			},
+
+			/**
+			 * Check for:
+			 *
+			 *     import X from 'react-dom'
+			 */
+			ImportDefaultSpecifier(node) {
+				if (isReactDOMImport(node)) {
+					add(foundNamespaces, context.getDeclaredVariables(node));
+				}
+			},
+
+			/**
+			 * Check for:
+			 *
+			 *     import * as X from 'react-dom'
+			 */
+			ImportNamespaceSpecifier(node) {
+				if (isReactDOMImport(node)) {
+					add(foundNamespaces, context.getDeclaredVariables(node));
+				}
+			},
+
+			/**
+			 * Check for:
+			 *
+			 *     import {x} from 'react-dom';
+			 *     import {x as y} from 'react-dom';
+			 */
+			ImportSpecifier(node) {
+				if (isReactDOMImport(node)) {
+					add(
+						foundBindings,
+						context
+							.getDeclaredVariables(node)
+							.filter((variable) => {
+								return (
+									variable.defs[0] &&
+									variable.defs[0].node &&
+									variable.defs[0].node.imported &&
+									variable.defs[0].node.imported.name ===
```

Comment:
Perhaps we can give a friendly name to the `variable.defs[0]` and that way store it only once?

Reply:
This is copied from `no-react-dom-render.js` so I think I will just leave it as is for now so that we don't have to re-write the other. This code is used to check for `import {x as y} from 'foo-bar';`. We take the definition (`y`) and check the name of its import (`x`). This way people can't circumnavigate defining something like `import {createPortal as fooCreatePortal} from 'react-dom';`